hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestOzoneConfigurationFields.java

```diff
@@ -65,6 +64,8 @@ public void initializeMemberVariables() {
     xmlPrefixToSkipCompare.add("ipc.client.rpc-timeout.ms");
     xmlPropsToSkipCompare.add("ozone.om.leader.election.minimum.timeout" +
         ".duration"); // Deprecated config
+    xmlPrefixToSkipCompare.add("ozone.s3g");
+    configurationPrefixToSkipCompare.add("ozone.s3g");
```

Comment:
All S3 config keys are ignored for the check?

Reply:
This one is a special sanity test it verifies that config properties keys are present in both XML and Java.  Ignoring all S3-specific keys is a potential source of mismatch.

This test is not needed for config defined with `@Config` annotations, so we may be able to alleviate the problem by moving from XML to the annotation-based approach.

Reply:
Yes, I believe that the check for properties to be in sync (xml and config keys) is done elsewhere as well for the s3 gateway making inclusion of "ozone.s3g" redundant.  Do you know where the other check is for the s3 gateway properties?

Reply:
@neils-dev Sorry, I missed this.

> I believe that the check for properties to be in sync (xml and config keys) is done elsewhere as well for the s3 gateway

I don't think there is such a check elsewhere.  `ozone-default.xml` defines properties for several components.  The test needs to have access to all Java config key definitions, that's why it's in the `ozone-integration-test` module.  A separate test would be possible only if config keys were defined in a separate XML file.

However, I don't see any problem with restoring the dependency and adding back `S3GatewayConfigKeys` in the test.  Ozone builds fine with the patch below.  Can you please let me know why the dependency would need to be removed?

```diff
diff --git hadoop-ozone/integration-test/pom.xml hadoop-ozone/integration-test/pom.xml
index d3faa6b52..cea5163fc 100644
--- hadoop-ozone/integration-test/pom.xml
+++ hadoop-ozone/integration-test/pom.xml
@@ -53,6 +53,10 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
       <scope>test</scope>
     </dependency>

+    <dependency>
+      <groupId>org.apache.ozone</groupId>
+      <artifactId>ozone-s3gateway</artifactId>
+    </dependency>
     <dependency>
       <groupId>org.apache.ozone</groupId>
       <artifactId>ozone-csi</artifactId>
diff --git hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestOzoneConfigurationFields.java hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestOzoneConfigurationFields.java
index 4d8151f78..3b9d3df97 100644
--- hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestOzoneConfigurationFields.java
+++ hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestOzoneConfigurationFields.java
@@ -26,6 +26,7 @@
 import org.apache.hadoop.ozone.om.OMConfigKeys;
 import org.apache.hadoop.hdds.scm.ScmConfigKeys;
 import org.apache.hadoop.ozone.recon.ReconServerConfigKeys;
+import org.apache.hadoop.ozone.s3.S3GatewayConfigKeys;

 import java.util.Arrays;
 import org.junit.Rule;
@@ -49,6 +50,7 @@ public void initializeMemberVariables() {
         new Class[] {OzoneConfigKeys.class, ScmConfigKeys.class,
             OMConfigKeys.class, HddsConfigKeys.class,
             ReconConfigKeys.class, ReconServerConfigKeys.class,
+            S3GatewayConfigKeys.class,
             SCMHTTPServerConfig.class,
             SCMHTTPServerConfig.ConfigStrings.class,
             ScmConfig.ConfigStrings.class
@@ -64,8 +66,6 @@ public void initializeMemberVariables() {
     xmlPrefixToSkipCompare.add("ipc.client.rpc-timeout.ms");
     xmlPropsToSkipCompare.add("ozone.om.leader.election.minimum.timeout" +
         ".duration"); // Deprecated config
-    xmlPrefixToSkipCompare.add("ozone.s3g");
-    configurationPrefixToSkipCompare.add("ozone.s3g");
     configurationPropsToSkipCompare
         .add(ScmConfig.ConfigStrings.HDDS_SCM_INIT_DEFAULT_LAYOUT_VERSION);
     // Currently replication and type configs moved to server side.
```

Reply:
Thanks for following up on removing `S3GatewayConfigKeys` with the associated integration test, TestOzoneConfigurationKeys @adoroszlai.  Doing so allows us to remove the `ozone-s3gateway` jar from the integration test classpath.
 
> Can you please let me know why the dependency would need to be removed?

The change was made to remove the `ozone-s3gateway `dependency from the integration test classpath (we can do away with including the `ozone-s3gateway` from the pom build file for the integration tests).  This was needed because initially with the `OmTransportFactory` selecting the `OmTransport` through the` ServiceProvider`.  In the case of the integration tests, the `ServiceProvider` was dynamically instantiating the `GrpcOmTransport` for **_all_** OzoneClients since it found the `ozone-s3gateway` in the classpath.

Reply:
Reading `OmTransportFactory` code, it should load dynamically via `ServiceLoader` only if `ozone.om.transport.class` is not set to the default one (`Hadoop3OmTransportFactory`).  I don't think integration tests would use `GrpcOmTransport`, unless we override this config in `hadoop-ozone/integration-test/src/test/resources/ozone-site.xml` or in individual test classes.

Reply:
In checking the integration tests and potential problems resulting from including the ozone-s3gateway in the integration tests classpaths, currently there are no problems.  However there is a case where it can cause problems.  The case where the service loader dynamically instantiates the GrpcOmTransport for integration tests currently occurs when the default `ozone.om.transport.class` config key is _mis-configured_ to `GrpcOmTransportFactory`.   

When the OmConfigKeys is incorrectly set to , OZONE_OM_TRANSPORT_CLASS_DEFAULT = "org.apache.hadoop.ozone.om.protocolPB".GrpcOmTransportFactory", it results in all OzoneClient requests using the GrpcOmTransport, failing tests such as `TestOzoneRpcClient`. 

 We can ensure that integration tests pass with the default config mis-configured by adding a` META-INF.services` for `ServiceProvider` to use hadoopOmTransport and overriding the _ozone.om.transport.class_ in the integration tests local _ozone-site.xml_ file.  Is this necessary or should we fail the integration tests if the _**default**_ _ozone.om.transport.class_ is set incorrectly in the OMConfigKeys?

We can have the `TestOzoneConfigurationKeys` include s3 gateway checks provided we ensure the tests pass if configuration keys are mis-configured or we simply fail if incorrectly set.

Reply:
@neils-dev Thanks for checking.

> should we fail the integration tests if the _**default**_ _ozone.om.transport.class_ is set incorrectly in the OMConfigKeys?

Yes, I think we should let them fail if there is a misconfiguration.  This provides a safety net.  So no IT-specific config necessary for now.